### ISSUE TITLE
chain restart of docker_service_manager_upstart resource

### ISSUE THREAD

**Issue by jperville** [chef-cookbooks/docker#643]
I tried upgrading docker cookbook from version 2.4.12 to version 2.4.20 and now my docker daemon is chain-restarting. I had to revert to version 2.4.12.

For information, I am using chef 12.4.3 on ubuntu 14.04 (64 bit) host and have some custom options for my docker_service. I wanted to try the cleaned up version of the upstart implementation because I prefer my options to be visible in `/etc/default/docker` DOCKER_OPTS instead of hardcoded in the upstart configuration.

Here is the relevant trace of the first chef-client after updating the docker cookbook to 2.4.20 :

```sh
Recipe: pm-docker::default
  * docker_service[default] action create
    * docker_installation_package[default] action create
      * apt_package[docker-engine] action install (up to date)
       (up to date)
     (up to date)
  * docker_service[default] action start
    * docker_service_manager_upstart[default] action start[2016-02-12T11:18:04+01:00] WARN: Default value false is invalid for property tls_ca_cert of resource . In Chef 13 this will become an error: Property tls_ca_cert must be one of: String, nil!  You passed false..

      * template[/etc/init/docker.conf] action create
        - update content in file /etc/init/docker.conf from 86549a to a9181d
        --- /etc/init/docker.conf	2016-02-11 10:59:34.304742175 +0100
        +++ /tmp/chef-rendered-template20160212-5207-1uvjil6	2016-02-12 11:18:04.793813209 +0100
        @@ -1,4 +1,4 @@
        -description "docker daemon"
        +description "Docker daemon"
         
         start on (filesystem and net-device-up IFACE!=lo)
         stop on runlevel [!2345]
        @@ -6,65 +6,56 @@
         limit nproc 524288 1048576
         
         respawn
        -respawn limit 2 10
         
        -pre-start script
        -  [ -x /sbin/sysctl ] || exit 6
        -  /sbin/sysctl -w net.ipv4.ip_forward=1 > /dev/null 2>&1
        -  /sbin/sysctl -w net.ipv6.conf.all.forwarding=1 > /dev/null 2>&1
        +kill timeout 20
         
        -  # see also https://github.com/tianon/cgroupfs-mount/blob/master/cgroupfs-mount
        -  if grep -v '^#' /etc/fstab | grep -q cgroup \
        -    || [ ! -e /proc/cgroups ] \
        -    || [ ! -d /sys/fs/cgroup ]; then
        -    exit 0
        -  fi
        -  if ! mountpoint -q /sys/fs/cgroup; then
        -    mount -t tmpfs -o uid=0,gid=0,mode=0755 cgroup /sys/fs/cgroup
        -  fi
        -  (
        -    cd /sys/fs/cgroup
        -    for sys in $(awk '!/^#/ { if ($4 == 1) print $1 }' /proc/cgroups); do
        -      mkdir -p $sys
        -      if ! mountpoint -q $sys; then
        -        if ! mount -n -t cgroup -o $sys cgroup $sys; then
        -          rmdir $sys || true
        -        fi
        -      fi
        -    done
        -  )
        +pre-start script
        +	# see also https://github.com/tianon/cgroupfs-mount/blob/master/cgroupfs-mount
        +	if grep -v '^#' /etc/fstab | grep -q cgroup \
        +		|| [ ! -e /proc/cgroups ] \
        +		|| [ ! -d /sys/fs/cgroup ]; then
        +		exit 0
        +	fi
        +	if ! mountpoint -q /sys/fs/cgroup; then
        +		mount -t tmpfs -o uid=0,gid=0,mode=0755 cgroup /sys/fs/cgroup
        +	fi
        +	(
        +		cd /sys/fs/cgroup
        +		for sys in $(awk '!/^#/ { if ($4 == 1) print $1 }' /proc/cgroups); do
        +			mkdir -p $sys
        +			if ! mountpoint -q $sys; then
        +				if ! mount -n -t cgroup -o $sys cgroup $sys; then
        +					rmdir $sys || true
        +				fi
        +			fi
        +		done
        +	)
         end script
         
         script
        -  # modify these in /etc/default/$UPSTART_JOB (/etc/default/docker)
        -  DOCKER=/usr/bin/$UPSTART_JOB
        -  DOCKER_DAEMON_OPTS=
        -  if [ -f /etc/default/$UPSTART_JOB ]; then
        -    . /etc/default/$UPSTART_JOB
        -  fi
        -  exec /usr/bin/docker daemon --bip=172.17.42.1/16 --group=docker --log-driver=json-file --pidfile=/var/run/docker.pid --userland-proxy=false >> /var/log/docker.log
        +	# modify these in /etc/default/$UPSTART_JOB (/etc/default/docker)
        +	DOCKER=/usr/bin/$UPSTART_JOB
        +	DOCKER_OPTS=
        +	if [ -f /etc/default/$UPSTART_JOB ]; then
        +		. /etc/default/$UPSTART_JOB
        +	fi
        +	exec "$DOCKER" daemon $DOCKER_OPTS
         end script
         
         # Don't emit "started" event until docker.sock is ready.
         # See https://github.com/docker/docker/issues/6647
         post-start script
        -    running() {
        -        RUNNING=1
        -        RESPONSE=`/usr/bin/docker  ps 2>&1 | head -n 1`
        -        local mret=$?
        -        if [ $mret -eq 0 ] \
        -        || [ `echo $RESPONSE | grep -qi "docker daemon"` ]; then
        -            RUNNING=0
        -        fi
        -        return $RUNNING
        -    }
        -
        -    while ! running ; do
        -      initctl status $UPSTART_JOB | grep -q "stop/" && exit 1
        -      echo "Waiting for Docker"
        -      sleep 0.1
        -    done
        -
        -    echo "Docker is up"
        +	DOCKER_OPTS=
        +	if [ -f /etc/default/$UPSTART_JOB ]; then
        +		. /etc/default/$UPSTART_JOB
        +	fi
        +	if ! printf "%s" "$DOCKER_OPTS" | grep -qE -e '-H|--host'; then
        +		while ! [ -e /var/run/docker.sock ]; do
        +			initctl status $UPSTART_JOB | grep -qE "(stop|respawn)/" && exit 1
        +			echo "Waiting for /var/run/docker.sock"
        +			sleep 0.1
        +		done
        +		echo "/var/run/docker.sock is up"
        +	fi
         end script
      * template[/etc/default/docker] action create
        - update content in file /etc/default/docker from 0a4467 to d081ab
        --- /etc/default/docker	2015-11-20 13:47:10.000000000 +0100
        +++ /tmp/chef-rendered-template20160212-5207-1ebao8z	2016-02-12 11:18:04.805813092 +0100
        @@ -1,14 +1,12 @@
         # Docker Upstart and SysVinit configuration file
         
         # Customize location of Docker binary (especially for development testing).
        -#DOCKER="/usr/local/bin/docker"
        +DOCKER="/usr/bin/docker"
         
         # Use DOCKER_OPTS to modify the daemon startup options.
        -#DOCKER_OPTS="--dns 8.8.8.8 --dns 8.8.4.4"
        +DOCKER_OPTS="--bip=172.17.42.1/16 --group=docker --log-driver=json-file --pidfile=/var/run/docker.pid --userland-proxy=false"
         
         # If you need Docker to use an HTTP proxy, it can also be specified here.
        -#export http_proxy="http://127.0.0.1:3128/"
         
         # This is also a handy place to tweak where Docker's temporary files go.
        -#export TMPDIR="/mnt/bigdrive/docker-tmp"
      * docker_service_manager_upstart[default] action restart
        * service[docker] action stop
          - stop service service[docker]
        * template[/etc/init/docker.conf] action create (up to date)
        * template[/etc/default/docker] action create (up to date)
        * service[docker] action start
          - start service service[docker]
        * bash[docker-wait-ready default] action run (skipped due to not_if)
      
      * docker_service_manager_upstart[default] action restart
        * service[docker] action stop
          - stop service service[docker]
        * template[/etc/init/docker.conf] action create (up to date)
        * template[/etc/default/docker] action create (up to date)
        * service[docker] action start
          - start service service[docker]
        * bash[docker-wait-ready default] action run (skipped due to not_if)
      
(snip, it chains restarts until I abort the chef-run)
      
      * docker_service_manager_upstart[default] action restart
        * service[docker] action stop^C[2016-02-12T11:19:21+01:00] FATAL: SIGINT received, stopping
[2016-02-12T11:19:21+01:00] FATAL: SIGINT received, stopping
root@staging4-backends-lb:/etc# ^C
          ================================================================================
          Error executing action `stop` on resource 'service[docker]'
          ================================================================================
          
          SystemExit
          ----------
          exit
          
          Cookbook Trace:
          ---------------
          /var/chef/cache/cookbooks/compat_resource/files/lib/chef_compat/copied_from_chef/chef/provider.rb:119:in `compile_and_converge_action'
          /var/chef/cache/cookbooks/docker/libraries/docker_service_manager_upstart.rb:68:in `block in <class:DockerServiceManagerUpstart>'
          /var/chef/cache/cookbooks/compat_resource/files/lib/chef_compat/copied_from_chef/chef/provider.rb:118:in `instance_eval'
          /var/chef/cache/cookbooks/compat_resource/files/lib/chef_compat/copied_from_chef/chef/provider.rb:118:in `compile_and_converge_action'
          /var/chef/cache/cookbooks/compat_resource/files/lib/chef_compat/copied_from_chef/chef/provider.rb:119:in `compile_and_converge_action'
          /var/chef/cache/cookbooks/compat_resource/files/lib/chef_compat/copied_from_chef/chef/provider.rb:119:in `compile_and_converge_action'
          
          Resource Declaration:
          ---------------------
          # In /var/chef/cache/cookbooks/docker/libraries/docker_service_manager_upstart.rb
          
           60:       service docker_name do
           61:         provider Chef::Provider::Service::Upstart
           62:         supports status: true
           63:         action :stop
           64:       end
           65:     end
          
          Compiled Resource:
          ------------------
          # Declared in /var/chef/cache/cookbooks/docker/libraries/docker_service_manager_upstart.rb:60:in `block in <class:DockerServiceManagerUpstart>'
          
          service("docker") do
```

**Comment by someara** [comment]
I tried and failed to reproduce this..
https://gist.github.com/someara/ab0bb3783b1e7f0126de

I spun up an Ubuntu-14.04 node, removed /etc/default/docker and re-ran chef-client.
Worked as expected... 

Perhaps it's somehow an upgrade path issue?

**Comment by jperville** [comment]
Hello @someara, and thanks for spending time on my issue.

I am reproducing the bug:
 - by running `kitchen converge service-upstart-ubuntu-1404`
 - after checking-out revision f833591b85915281b1bcb186976c33509e4482de of this cookbook (current master HEAD as of today 12th February 2016)
 - and modified .kitchen.yml to set `require_chef_omnibus: 12.4.3` instead of `require_chef_omnibus: true`

See my gist with chef-client 12.4.3 : https://gist.github.com/jperville/0b0dd108b65524fbe114

Please note that your gist above is using latest chef-client (version 12.7.2) while I am using (and have to stay for now since have at least one legacy cookbook subtlety broken with chef 2.5.x) on version 12.4.3. This may explain the difference because I ran the test suite again using chef-client 12.7.2 and was able to converge successfully.

I have done some further checking with the different supported chef-versions:
 - chef-client 2.7.2 : works
 - chef-client 2.6.0 : works
 - chef-client 2.5.1 : works
 - chef-client 2.4.3 : broken

If you have a way to restore compatibility with my chef version I would be very grateful.

**Comment by someara** [comment]
Using the delayed notification timing seems to have fixed this up.
Fix released in 2.4.24.

**Comment by jperville** [comment]
Cheers @someara .
